File: src/main/java/com/trilead/ssh2/crypto/keys/Ed25519KeyFactory.java
package com.trilead.ssh2.crypto.keys;

import java.security.InvalidKeyException;
import java.security.Key;
import java.security.KeyFactorySpi;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.KeySpec;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;

public class Ed25519KeyFactory extends KeyFactorySpi {
	@Override
	protected PublicKey engineGeneratePublic(KeySpec keySpec) throws InvalidKeySpecException {
		if (keySpec instanceof X509EncodedKeySpec) {
			return new Ed25519PublicKey((X509EncodedKeySpec) keySpec);
		}
		throw new InvalidKeySpecException("Unrecognized key spec: " + keySpec.getClass());
	}

	@Override
	protected PrivateKey engineGeneratePrivate(KeySpec keySpec) throws InvalidKeySpecException {
		if (keySpec instanceof PKCS8EncodedKeySpec) {
			return new Ed25519PrivateKey((PKCS8EncodedKeySpec) keySpec);
		}
		throw new InvalidKeySpecException("Unrecognized key spec: " + keySpec.getClass());
	}

	@Override
	protected <T extends KeySpec> T engineGetKeySpec(Key key, Class<T> keySpec) throws InvalidKeySpecException {
		throw new InvalidKeySpecException("not implemented yet " + key + " " + keySpec);
	}

	@Override
	public Key engineTranslateKey(Key key) throws InvalidKeyException {
		if (key instanceof Ed25519PublicKey || key instanceof Ed25519PrivateKey) {
			return key;
		}
		if (key instanceof PublicKey && key.getFormat().equals("X.509")) {
			byte[] encoded = key.getEncoded();
			try {
				return new Ed25519PublicKey(new X509EncodedKeySpec(encoded));
			} catch (InvalidKeySpecException e) {
				throw new InvalidKeyException(e);
			}
		}

		if (key instanceof PrivateKey && key.getFormat().equals("PKCS#8")) {
			byte[] encoded = key.getEncoded();
			try {
				return new Ed25519PrivateKey(new PKCS8EncodedKeySpec(encoded));
			} catch (InvalidKeySpecException e) {
				throw new InvalidKeyException(e);
			}
		}

		throw new InvalidKeyException("Could not convert EdDSA key from key class " + key.getClass());
	}
}

File: sshlib/src/main/kotlin/org/connectbot/sshlib/crypto/ed25519/Ed25519KeyFactory.kt
/*
 * Copyright 2025 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.sshlib.crypto.ed25519

import java.security.InvalidKeyException
import java.security.Key
import java.security.KeyFactorySpi
import java.security.PrivateKey
import java.security.PublicKey
import java.security.spec.InvalidKeySpecException
import java.security.spec.KeySpec
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec

internal class Ed25519KeyFactory : KeyFactorySpi() {
    override fun engineGeneratePublic(keySpec: KeySpec): PublicKey {
        if (keySpec is X509EncodedKeySpec) {
            return Ed25519PublicKey(keySpec)
        }
        throw InvalidKeySpecException("Unrecognized key spec: ${keySpec.javaClass}")
    }

    override fun engineGeneratePrivate(keySpec: KeySpec): PrivateKey {
        if (keySpec is PKCS8EncodedKeySpec) {
            return Ed25519PrivateKey(keySpec)
        }
        throw InvalidKeySpecException("Unrecognized key spec: ${keySpec.javaClass}")
    }

    override fun <T : KeySpec> engineGetKeySpec(key: Key, keySpec: Class<T>): T = throw InvalidKeySpecException("Not implemented: $key $keySpec")

    override fun engineTranslateKey(key: Key): Key {
        if (key is Ed25519PublicKey || key is Ed25519PrivateKey) {
            return key
        }

        if (key is PublicKey && key.format == "X.509") {
            try {
                return Ed25519PublicKey(X509EncodedKeySpec(key.encoded))
            } catch (e: InvalidKeySpecException) {
                throw InvalidKeyException(e)
            }
        }

        if (key is PrivateKey && key.format == "PKCS#8") {
            try {
                return Ed25519PrivateKey(PKCS8EncodedKeySpec(key.encoded))
            } catch (e: InvalidKeySpecException) {
                throw InvalidKeyException(e)
            }
        }

        throw InvalidKeyException("Could not convert EdDSA key from key class ${key.javaClass}")
    }
}
